torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -308,65 +260,47 @@ PyObject* THPVariable_getitem(PyObject* self, PyObject* index) {
   HANDLE_TH_ERRORS
   auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
   OptionalDeviceGuard device_guard(device_of(self_));
+  bool is_tracing = torch::jit::tracer::isTracing();
 
-  // handle simple types: integers, slices, ellipsis
-  if (index == Py_None) {
-    return wrap(self_.unsqueeze(0));
-  } else if (index == Py_Ellipsis) {
-    return wrap(at::alias(self_));
-  } else if (THPUtils_checkLong(index)) {
-    return wrap(applySelect(self_, 0, index));
-  } else if (PySlice_Check(index)) {
-    return wrap(applySlice(self_, 0, index, true));
+  // handle simple types: integers, slices, ellipsis, none
+  if (THPUtils_checkLong(index) \
+    || PySlice_Check(index) \
+    || index == Py_Ellipsis \
+    || index == Py_None) {
+    return THPVariable_Wrap(at::indexing::handleSimpleTypesInSingleDimIndexingGet(
+      self_,
+      traceAndConvertPythonIndexToTensorIndex(self_, index, is_tracing),
+      /*is_tracing=*/is_tracing));
```

Comment:
I'm not going to comment on whether or not the slow down is "acceptable" or not. Instead, I want to ask a question: are we just doing dumb shit which is causing performance to be bad, and can we (easily) stop hurting ourselves and recover the performance.

Here, I see a fairly salient case where there is "dumb stuff" going on that is likely making things worse: you are doing Python C API tests twice in the hotpath: at the top level you test for `THPUtils_checkLong` and `PySlice_Check` here, and then inside `handleSimpleTypesInSingleDimIndexingGet` you test them again. If the compiler is not able to tell that `PySlice_Check` is a pure function (pretty unlikely, as I expect `PySlice_Check` is not getting inlined), it now has to do this test twice, for no good reason. You could verify this fairly simply by inspecting the generated assembly on these functions.

So what should you do in this case? I can think of two approaches:

1. The use of `traceAndConvertPythonIndexToTensorIndex` helper function really is not buying you much. The modified code here is exactly the same number of lines as the old code. It would be better to just replicate the logic here (try for something very similar what you saw before), and just chalk this up as an "optimized fast path". This makes sense: in principle, you can delete these lines entirely and you'll end up with a correct implementation, because the "advanced" tuple handling can handle 1-element tuples. You'll need a way to both trace and apply slice, but this can probably be done via minor refactoring of the helper functions in this file.

2. You could unconditionally run `traceAndConvertPythonIndexToTensorIndex` here, and return an `optional<TensorIndex>` instead of a `TensorIndex`. The nullopt indicates that you are in the tuply case, in which case you fall through to the advanced index handling.